crates/wiggle/src/lib.rs

```diff
@@ -731,6 +732,24 @@ impl<'a> GuestPtr<'a, str> {
             UnsafeBorrowResult::Err(e) => Err(e),
         }
     }
+
+    /// Attempts to create a [`GuestStrCow<'_>`] from this pointer, performing
+    /// bounds checks and utf-8 checks. Whereas [`GuestPtr::as_str`] will fail
+    /// with `None` if attempting to access Wasm shared memory, this call will
+    /// succeed: if used on shared memory, this function will copy the string
+    /// into [`GuestStrCow::Copied`]. If the memory is non-shared, this returns
+    /// a [`GuestStrCow::Borrowed`] (a thin wrapper over [`GuestStr<'_, T>]`).
+    pub fn as_cow(&self) -> Result<GuestStrCow<'a>, GuestError> {
+        match self.as_bytes().as_unsafe_slice_mut()?.shared_borrow() {
+            UnsafeBorrowResult::Ok(s) => Ok(GuestStrCow::Borrowed(GuestStr(s))),
```

Comment:
I think this may need a `.try_into()` or otherwise it's skipping the utf-8 check?

Reply:
Ah, good catch! Rebased with that additional change as a follow-on commit.